pallets/pallet-deposit-storage/src/try_state.rs

```diff
@@ -0,0 +1,91 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2024 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+//! Pallet to store namespaced deposits for the configured `Currency`. It allows
+//! the original payer of a deposit to claim it back, triggering a hook to
+//! optionally perform related actions somewhere else in the runtime.
+//! Each deposit is identified by a namespace and a key. There cannot be two
+//! equal keys under the same namespace, but the same key can be present under
+//! different namespaces.
+
+use frame_support::{ensure, traits::fungible::InspectHold};
+use frame_system::pallet_prelude::BlockNumberFor;
+use kilt_support::Deposit;
+use parity_scale_codec::{Decode, Encode};
+use sp_runtime::{traits::CheckedAdd, TryRuntimeError};
+use sp_std::collections::btree_map::BTreeMap;
+
+use crate::{deposit::DepositEntry, AccountIdOf, BalanceOf, Config, Deposits, HoldReason};
+
+pub(crate) fn try_state<T>(n: BlockNumberFor<T>) -> Result<(), TryRuntimeError>
+where
+	T: Config,
+{
+	crate::fungible::try_state::check_fungible_consistency::<T>(n)?;
+	check_regular_deposits_consistency::<T>(n)?;
+
+	Ok(())
+}
+
+// Verify the state outside of the `MutateHold` implementation does not
+// interfere with the `MutateHold` state.
+fn check_regular_deposits_consistency<T>(_n: BlockNumberFor<T>) -> Result<(), TryRuntimeError>
+where
+	T: Config,
+{
+	// Sum together all the deposits stored not part of the `MutateHold`
+	// implementation, and fail if any of them has the unexpected
+	// `crate::HoldReason::FungibleImpl` reason.
+	let sum_of_deposits = Deposits::<T>::iter_values().try_fold(
+		// We can't use `T::RuntimeHoldReason` as a key because it does not implement `Ord`, so we `.encode()` it here.
+		BTreeMap::<(AccountIdOf<T>, Vec<u8>), BalanceOf<T>>::new(),
+		|mut sum,
+		 DepositEntry {
+		     reason,
+		     deposit: Deposit { amount, owner },
+		 }| {
+			// Regular deposits should not interfere with the `MutateHold` implementation
+			// state.
+			ensure!(
+				reason != HoldReason::FungibleImpl.into(),
```

Comment:
I think, if we ever add a new `Reason` we still want to only check the Deposit reason right? maybe we can make this check explicit.

Reply:
Sorry, I am not sure I understand your point?

Reply:
`Reason == HoldReason::FigibleImpl` or is this not intendant? If yes, what if a new Reason is introduced?

Reply:
We check for `reason != HoldReason::FungibleImpl.into(),`, at least in this line, so I still don't get your point. Here we're only interested in checking that `HoldReason::FungibleImpl` is only used for the trait implementation. If there's a new deposit reason, we don't care.

Reply:
Okay I see